### ISSUE TITLE
Add validation during training

### ISSUE THREAD

**Issue by jrao1** [ibab/tensorflow-wavenet#123]
I think the paper mentioned using validation dataset, so this may be something worth adding. 

This may require some changes on AudioReader. Right now it's hardcoded to load all files in a directory, looks to me it would be better to give it a generator, and move load_generic_audio and similar to train.py since they're training details.

**Comment by ibab** [comment]
My current goal is to refactor everything that's specific to certain datasets into `AudioReader`s so that the rest of the implementation can be more generic.
(And the `train.py` and `generate.py` scripts should become as small as possible, as they're tricky to test properly)
I can imagine that we could give the `AudioReader` a method to get a train and a test batch, like this:

```python
reader = MyAudioReader('./directory')
X = reader.train_batch()
X_test = reader.train_batch()
```

`X` and `X_test` would be TensorFlow ops that we can pass to the network.

**Comment by jrao1** [comment]
So the idea is to have a AudioReader superclass, and allow extension by subclassing?

**Comment by ibab** [comment]
@jrao1: Yeah, `AudioReader` can define the interface that's expected by the model.

**Comment by jrao1** [comment]
@ibab I have a preliminary version here: https://github.com/jrao1/tensorflow-wavenet/tree/add_validation, let me know this fits your ideas.

**Comment by lukovkin** [comment]
@jrao1 Thank you for your implementation, but when I tried to use it (but implementing another class instead of DirectoryAudioReader for another type of data), it looked like the execution was stuck within wavenet.audio_reader.AudioReader#thread_main, not actually reaching the training phase.
I've added an output of processed filenames and it repeats names of training and validation files all over again.

Would you be so kind to help to overcome this issue, because I'm not very much acquainted to  the TF queues and threading.

**Comment by jrao1** [comment]
@lukovkin Could you provide a way to reproduce this issue?

**Comment by lukovkin** [comment]
@jrao1 Yes, I made a branch which have code closest to your initial implementation (as I believe) - https://github.com/DeepGnosis/tensorflow-wavenet.git#jrao1

Command line I've used:
python train.py --data_dir=data --sample_size 50000 --val_size 3329

I could provide sample train and validation files also.

**Comment by jrao1** [comment]
@lukovkin This is probably because your sample size is bigger than the data size from the file, it's sort of a bug, the logic under `if sample_size:` in method `get_audio_iterator` assumes sample_size is always smaller than the size of the audio, so it needs to cut audio into pieces, it never considered the case where sample_size is bigger than audio, in which case `yield` never got executed. You can add a line:

`print 'train', train, 'sample_size', sample_size, 'audio size', audio.shape, 'buffer size', buffer_.shape`

under `if sample_size:` to verify this

**Comment by lukovkin** [comment]
@jrao1 Looks like you're quite right!
Actually, in my case the data size was exactly equal to the sample size, but the loop condition was '>'.
I've changed to:
`while len(buffer_) >= sample_size:`
and it looks to be working.

Thank you.

**Comment by jrao1** [comment]
@lukovkin No problem, glad i could help.
